File: hackerrank/30 days of code/day 4, class vs instance/class_vs_instance.cpp
// https://www.hackerrank.com/challenges/30-class-vs-instance/problem?isFullScreen=true
// From C++23 onwards

#include <iostream>
#include <print> // In C++23

using namespace std;

class Person {
    int age;

public:
    explicit Person(int initial_age);
    void am_i_old() const;
    void year_passes();
};

Person::Person(const int initial_age)
{
    if (initial_age < 0) {
        age = 0;
        cout << "Age is not valid, setting age to 0.\n";
    } else
        age = initial_age;
}

void Person::am_i_old() const
{
    if (age < 13)
        cout << "You are young.\n";
    else if (age >= 13 && age < 18)
        cout << "You are a teenager.\n";
    else
        cout << "You are old.\n";
}

void Person::year_passes()
{
    ++age;
}

int main()
{
    int n_tests;
    cin >> n_tests;

    for (int age; cin >> age && n_tests-- > 0;) {
        Person p{age};
        p.am_i_old();

        for (int i{3}; i--; p.year_passes());
        p.am_i_old();

        println("");
    }

    return 0;
}

File: hackerrank/30 days of code/day 4, class vs instance/class_vs_instance.rs
// https://www.hackerrank.com/challenges/30-class-vs-instance/problem?isFullScreen=true

use text_io::read;

struct Person {
    age: i32,
}

impl Person {
    fn new(mut initial_age: i32) -> Person {
        if initial_age < 0 {
            initial_age = 0;
            println!("Age is not valid, setting age to 0.");
        }
        return Person { age: initial_age };
    }

    fn am_i_old(&self) {
        if self.age < 13 {
            println!("You are young.");
        } else if self.age >= 13 && self.age < 18 {
            println!("You are a teenager.");
        } else {
            println!("You are old.");
        }
    }

    fn year_passes(&mut self) {
        self.age += 1;
    }
}

fn main() {
    let n_tests: i32 = read!();

    for _ in 0..n_tests {
        let age = read!();
        let mut p = Person::new(age);
        p.am_i_old();

        for _ in 0..3 {
            p.year_passes();
        }
        p.am_i_old();

        println!("");
    }
}
